### ISSUE TITLE
System.TypeLoadException after upgrade to 6.0.0-rc.1.21452.10 : Method 'AppendIdentityWhereCondition' does not have an implementation

### ISSUE THREAD

**Issue by died** [dotnet/efcore#26059]
## File a bug

Was using EF Core 6.0 Preview 7 before, everything working normal, after upgrade to RC1, it throw exception on system startup.
```
Method 'AppendIdentityWhereCondition' in type 'Pomelo.EntityFrameworkCore.MySql.Update.Internal.MySqlUpdateSqlGenerator' from assembly 'Pomelo.EntityFrameworkCore.MySql, Version=6.0.0.0, Culture=neutral, PublicKeyToken=2cc498582444921b' does not have an implementation.
```

### Include your code

![img](https://i.imgur.com/rKcZmuQ.png)

### Include verbose output

Please include `--verbose` output when filing bugs about the `dotnet ef` or Package Manager Console tools.

```
PS C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot> dotnet ef dbcontext list --verbose
Using project 'C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\Coscomms.CryptoBot.csproj'.
Using startup project 'C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\Coscomms.CryptoBot.csproj'.
Writing 'C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\obj\Coscomms.CryptoBot.csproj.EntityFrameworkCore.targets'...
dotnet msbuild /target:GetEFProjectMetadata /property:EFProjectMetadataFile=C:\Users\Died\AppData\Local\Temp\tmp1D88.tmp /verbosity:quiet /nologo C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\Coscomms.CryptoBot.csproj
Writing 'C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\obj\Coscomms.CryptoBot.csproj.EntityFrameworkCore.targets'...
dotnet msbuild /target:GetEFProjectMetadata /property:EFProjectMetadataFile=C:\Users\Died\AppData\Local\Temp\tmp1F9C.tmp /verbosity:quiet /nologo C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\Coscomms.CryptoBot.csproj
Build started...
dotnet build C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\Coscomms.CryptoBot.csproj /verbosity:quiet /nologo

Build succeeded.
    0 Warning(s)
    0 Error(s)
Time Elapsed 00:00:02.32
Build succeeded.
dotnet exec --depsfile C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\bin\Debug\net6.0\Coscomms.CryptoBot.deps.json --additionalprobingpath C:\Users\Died\.nuget\packages --additionalprobingpath "D:\Program Files (x86)\Microsoft Visual Studio\Shared\NuGetPackages" --runtimeconfig C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\bin\Debug\net6.0\Coscomms.CryptoBot.runtimeconfig.json C:\Users\Died\.dotnet\tools\.store\dotnet-ef\6.0.0-rc.1.21452.10\dotnet-ef\6.0.0-rc.1.21452.10\tools\netcoreapp3.1\any\tools\netcoreapp2.0\any\ef.dll dbcontext list --assembly C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\bin\Debug\net6.0\Coscomms.CryptoBot.dll --project C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\Coscomms.CryptoBot.csproj --startup-assembly C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\bin\Debug\net6.0\Coscomms.CryptoBot.dll --startup-project C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\Coscomms.CryptoBot.csproj --project-dir C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\ --root-namespace Coscomms.CryptoBot --language C# --framework net6.0 --nullable --working-dir C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot --verbose
Using assembly 'Coscomms.CryptoBot'.
Using startup assembly 'Coscomms.CryptoBot'.
Using application base 'C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\bin\Debug\net6.0'.
Using working directory 'C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot'.
Using root namespace 'Coscomms.CryptoBot'.
Using project directory 'C:\Users\Died\source\repos\Coscomms\Coscomms.CryptoBot\'.
Remaining arguments: .
Finding DbContext classes...
Finding IDesignTimeDbContextFactory implementations...
Finding application service provider in assembly 'Coscomms.CryptoBot'...
Finding Microsoft.Extensions.Hosting service provider...
Using environment 'Development'.
Using application service provider from Microsoft.Extensions.Hosting.
Found DbContext 'CryptoBotContext'.
Finding DbContext classes in the project...
Coscomms.CryptoBotDb.CryptoBotContext
```

### Include provider and version information

EF Core version: Microsoft.EntityFrameworkCore 6.0.0-rc.1.21452.10
Database provider: Pomelo.EntityFrameworkCore.MySql 6.0.0-preview.5
Target framework:  .NET 6.0
Operating system: Win 10 Home Version 10.0.19043
IDE: Visual Studio 2022 Preview

**Comment by roji** [comment]
Duplicate of #3976

**Comment by died** [comment]
> @died you cannot mix versions - you will have to wait for Pomelo.EntityFrameworkCore.MySql to release an rc.1 version.

but Pomelo.EntityFrameworkCore.MySql preview 5 with Microsoft.EntityFrameworkCore preview 7 worked.

Or maybe EF core can Highlight not work with MySql since Pomelo didn't follow the the version ?

**Comment by roji** [comment]
> but Pomelo.EntityFrameworkCore.MySql preview 5 with Microsoft.EntityFrameworkCore preview 7 worked.

Version mixing **may** work if no provider-facing APIs change on the EF side, but may not. The only way that's guaranteed to work is to use matching versions, which means the provider was built and tested against the corresponding version.

**Comment by died** [comment]
> > but Pomelo.EntityFrameworkCore.MySql preview 5 with Microsoft.EntityFrameworkCore preview 7 worked.
> 
> Version mixing **may** work if no provider-facing APIs change on the EF side, but may not. The only way that's guaranteed to work is to use matching versions, which means the provider was built and tested against the corresponding version.

Just recommend, maybe it should list on breaking change or known issue. 
***EF core 6 RC1 didn't work with MySQL for now***

It have save many people half day to debug.

**Comment by XYCaptain** [comment]
TypeLoadException: Method 'AppendIdentityWhereCondition' in type 'Pomelo.EntityFrameworkCore.MySql.Update.Internal.MySqlUpdateSqlGenerator' from assembly 'Pomelo.EntityFrameworkCore.MySql, Version=6.0.0.0, Culture=neutral, PublicKeyToken=2cc498582444921b' does not have an implementation.
=.=

**Comment by roji** [comment]
@died and others: there is no breaking change, we simply do not support mixing different package versions, that's all.

**Comment by died** [comment]
> TypeLoadException: Method 'AppendIdentityWhereCondition' in type 'Pomelo.EntityFrameworkCore.MySql.Update.Internal.MySqlUpdateSqlGenerator' from assembly 'Pomelo.EntityFrameworkCore.MySql, Version=6.0.0.0, Culture=neutral, PublicKeyToken=2cc498582444921b' does not have an implementation.
> =.=

Uninstall RC1 , clean up, install Preview 7,  then it worked.

**Comment by XYCaptain** [comment]
> > TypeLoadException: Method 'AppendIdentityWhereCondition' in type 'Pomelo.EntityFrameworkCore.MySql.Update.Internal.MySqlUpdateSqlGenerator' from assembly 'Pomelo.EntityFrameworkCore.MySql, Version=6.0.0.0, Culture=neutral, PublicKeyToken=2cc498582444921b' does not have an implementation.
> > =.=
> 
> Uninstall RC1 , clean up, install Preview 7, then it worked.

Thanks.

**Comment by died** [comment]
> @died and others: there is no breaking change, we simply do not support mixing different package versions, that's all.

@roji Yes, Thank for the **"no breaking change"** breaking things, good job.

**Comment by ErikEJ** [comment]
@roji "Mixing different package versions" - should only be an issue until RTM?

**Comment by roji** [comment]
Guys, don't forget we are talking about prerelease previews - these are snapshots of ongoing work, where breaks sometimes need to happen (especially in the provider-facing APIs). Changes between previews are by definition not breaking changes, since nothing has been released yet. Asking the team to start publishing breaking change notes on a monthly basis between previews would take away valuable time that should be used for actual development.

> @roji "Mixing different package versions" - should only be an issue until RTM?

@ErikEJ how would that work? Breaking changes are gradually introduced as we do them, during the development cycle. At the moment when they're introduced, it becomes impossible to mix versions; "delaying" the breaking change to RTM and keeping all previews compatible with each other would mean that we only do breaking changes after rc2?

To summarize, @died I really am sorry you lost valuable time on this, and we may want to make it more visible in [the docs](https://docs.microsoft.com/en-us/ef/core/what-is-new/#get-it-now) that package mixing simply isn't supported. But other than that, does it really seem unreasonable to require users to wait for each preview release for their provider, and always use aligned versions?

**Comment by ErikEJ** [comment]
@roji I think you misunderstood, I mean package mixing after RTM.

**Comment by died** [comment]
@roji The issue is not hard, from the beginning I know it cause by RC1, rollback will solve it.
In my opinions, efcore RC1 will need method 'AppendIdentityWhereCondition' when initial so if provider didn't implement it, it won't work. It was a breaking change from Preview.

Yes you don't need post every change since it was prerelease version, but if someone facing the issue cause by the breaking change, you should explain it, not just throw the duty to third party provider.
![img](https://i.imgur.com/43nQOgT.png)
Especially you are paid by @microsoft and a member of @dotnet (or because you got paid so no responsibility ? whatever)

I think for normal unpaid from @microsoft user just want to know how to fix the exception, they don't need a official answer said "**NOT MY BUSINESS**".

You can just tell them rollback to preview version if don't want explain it. or explain it if you want to.

It can easily save you and other many people's time.

**Comment by roji** [comment]
@ErikEJ we sometimes get issues where users are trying to mix different major/minor versions too (e.g. 3.1 and 5.0, for various reasons). Basically it's never really supported to mix versions, except in patch versions of the same major/minor version, where it's almost guaranteed that the provider-facing API wouldn't break.

**Comment by roji** [comment]
@died,

>  In my opinions, efcore RC1 will need method 'AppendIdentityWhereCondition' when initial so if provider didn't implement it, it won't work.

Providers indeed need to sync up as EF Core changes; the change was intentional, and we work with provider writers to help them with this. There's simply no way around this - asking EF to never break providers would be the same as freezing EF and never allowing it to introduce new features.

> It was a breaking change from Preview.

I understand and acknowledge that it this change caused your application to break when upgrading from preview7 to rc1, and this is indeed a (provider-facing) breaking change between EF Core 5.0 and 6.0. But as preview7 is a preview and not a release, there is no breaking change here.

Aside from this, as I wrote above, I agree we could make it clearer in the docs that package mixing is not supported. Is there anything else you're expecting or suggesting we do here?

**Comment by died** [comment]
@roji Maybe something like that : [ASP.NET Core updates in .NET 6 Release Candidate 1](https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-6-rc-1/)
At least end user can know what they will facing, not try and error.
